src/Components/Components/src/RenderTree/Renderer.cs

```diff
@@ -199,53 +200,67 @@ protected Task RenderRootComponentAsync(int componentId)
 
-            // During the rendering process we keep a list of components performing work in _pendingTasks.
-            // _renderer.AddToPendingTasks will be called by ComponentState.SetDirectParameters to add the
-            // the Task produced by Component.SetParametersAsync to _pendingTasks in order to track the
-            // remaining work.
-            // During the synchronous rendering process we don't wait for the pending asynchronous
-            // work to finish as it will simply trigger new renders that will be handled afterwards.
-            // During the asynchronous rendering process we want to wait up until all components have
-            // finished rendering so that we can produce the complete output.
             var componentState = GetRequiredComponentState(componentId);
             if (HotReloadFeature.IsSupported)
             {
-                // when we're doing hot-reload, stash away the parameters used while rendering root components.
+                // When we're doing hot-reload, stash away the parameters used while rendering root components.
                 // We'll use this to trigger re-renders on hot reload updates.
-                _rootComponents ??= new();
-                _rootComponents.Add((componentState, initialParameters.Clone()));
+                _rootComponentsLatestParameters ??= new();
+                _rootComponentsLatestParameters[componentId] = initialParameters.Clone();
             }
 
             componentState.SetDirectParameters(initialParameters);
 
-            try
-            {
-                await ProcessAsynchronousWork();
-                Debug.Assert(_pendingTasks.Count == 0);
-            }
-            finally
+            await WaitForQuiescence();
+            Debug.Assert(_pendingTasks == null);
+        }
+
+        /// <summary>
+        /// Removes the specified component from the renderer, causing the component and its
+        /// descendants to be disposed.
+        /// </summary>
+        /// <param name="componentId">The ID of the root component.</param>
+        protected void RemoveRootComponent(int componentId)
+        {
+            Dispatcher.AssertAccess();
+
+            var rootComponentState = GetRequiredComponentState(componentId);
+            if (rootComponentState.ParentComponentState is not null)
             {
-                _pendingTasks = null;
+                throw new InvalidOperationException("The specified component is not a root component");
             }
+
+            // This assumes there isn't currently a batch in progress, and will throw if there is.
+            // Currently there's no known scenario where we need to support calling RemoveRootComponentAsync
+            // during a batch, but if a scenario emerges we can add support.
+            _batchBuilder.ComponentDisposalQueue.Enqueue(componentId);
+            _rootComponentsLatestParameters?.Remove(componentId);
+            ProcessRenderQueue();
```

Comment:
Will this trigger the disposal of all the components in the tree? I remember having to do something more "sophisticated" here (triggering an empty render).

Reply:
It does - disposal is recursive. There's a unit test for that too.

Reply:
Cool, I'm glad we found a simpler way, the way I did it was a bit convoluted